### PR TITLE
[web-console] Implement pipeline tags UI

### PR DESCRIPTION
**Pull request by Karakatiza666**
Testing: manual, added fda and python SDK tests, and web-console component test

I implemented the tag colors using a `|RRGGBB` suffix for the tag name. pipeline-manager stores tags as opaque strings, clients ensure tags are unique by the text label, not the entire text string.

Web-console can batch-update the pipelines (as per design) - so a tag can be deleted from all pipelines, and re-named and re-colored across all pipelines (one HTTP request per pipeline), in batches of 20 to avoid spamming too many HTTP calls at once.

The pipeline-manager allows duplication of colors for the same tag text label, so in the case this occurs web-console handles it gracefully and the user does not experience issues - they can just delete/re-color the offending color variant.

[Screencast from 2026-06-23 17-52-21.webm](https://github.com/user-attachments/assets/367cb173-f007-4556-a8de-4be07213108d)

### CODE REVIEW

**Review by mythical-fred — COMMENTED** [review]
Draft, so high-level only.

The shape of the UI is fine — `SlidingPanels` extraction from `DownloadSupportBundle` is a clean reuse; the list/create carousel inside `Tags.svelte` follows the same pattern. Optimistic update + `discardPendingListRefresh` + server PATCH for tag edits is the right pattern.

One architectural concern worth raising before this is marked ready:

**Encoding color as trailing-space padding on the tag string.** `tags.ts` notes the trade-off explicitly:

> A consequence is that two tags with the same visible text but different padding (`"foo"` vs `"foo "`) are genuinely different tags with different colors; the UI keeps them distinct rather than collapsing them.

This is a clever workaround for "backend doesn't know about color," but it leaks in places we can't always control:

1. Anything outside the web-console that touches `tags` — `fda`, the Python SDK, raw `PATCH /pipelines/:name`, programmatic creation in tests — sees an opaque whitespace-bearing string. A user who types `--tag prod` in the CLI and expects to land on the same colored chip they made in the UI will not.
2. Filtering / search / equality / set operations on tags have to be "padded equality." `matchesSearch` happens to do `trim().toLowerCase()` which is fine in isolation, but every future feature that touches tags has to remember the encoding.
3. Any HTTP/proxy layer or backend handler that normalizes whitespace (a `TRIM()` in SQL, a `.strip()` in Python, a hand-written validator that rejects trailing spaces) silently breaks color. The encoding is invisible from the server side, so the server can't even warn about it.
4. The "+0 padding" tag (length already lands on Red by accident) is indistinguishable in storage from a tag the user genuinely meant to leave at the default — so re-color is one-way OK but the round-trip semantics are subtle.

Better long-term shapes:

- **Backend stores `(tag, color)`** as a structured value (`{name, color_index}` or `name#color_index` with a documented separator that is not whitespace). Backend doesn't have to interpret color, but the field is named and the encoding is explicit.
- **Or: color is purely client-side**, derived from a hash of the tag name (you almost already do this — `tag.length % palette` is just a very weak hash). Drop the "user picks color" feature; let the palette be deterministic. This is what GitHub labels-from-strings does, and it removes the whole encoding problem.
- **Or: store color in localStorage / user prefs**, keyed by tag display name. Loses cross-user color consistency, gains a clean tag string.

If you have a specific reason for tying color to the persisted string (e.g. cross-user consistency without a backend round-trip for color), say so in the description so it's a deliberate choice; otherwise I'd avoid landing the padding trick.

Smaller things, save for full review when this is ready:

- `tagWithColor` builds `padEnd(base.length + padding)` — fine, but the math `((c - bL) + P) % P` is the kind of "trust me" arithmetic that screams for two assertions (`padding < paletteSize`, `(base.length + padding) % paletteSize === colorIndex`) or a unit test. Currently neither exists.
- `Tags.svelte` `setTags` fires `api.patchPipeline(...)` but does not await it or `.catch` it. If the PATCH fails the optimistic update is never reconciled — `discardPendingListRefresh()` is called eagerly, so the next list refresh won't immediately overwrite. Decide whether failures should toast or roll back.
- `SlidingPanels` is a nice extraction. Consider moving it into `common-ui` alongside `Tooltip`/`Popup` so the next caller doesn't import it from `pipelines/`-adjacent paths.

Not blocking anything since it's a draft; the encoding question is the only one I'd want resolved before ready-for-review.

File: python/feldera/rest/feldera_client.py [hunk 1]
```diff
@@ -398,6 +399,10 @@ def patch_pipeline(
         stored one, and a field left as ``None`` is omitted from the request and
         leaves the stored value untouched. To clear the description or tags, pass
         an empty string or empty list, respectively.
+
+        Tags, when provided, are normalized before being sent: color variants of
+        the same display name are collapsed and the result is stored in
```
Comment (mihaibudiu) [inline python/feldera/rest/feldera_client.py]: where is it stored?
or does this mean "represented"?

File: python/feldera/tags.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+``"prod|22c55e"`` are considered a tag collision which can but should not occur.
+
+To stay consistent with the web console, the SDK treats a tag's *display name* --
+its name with any color suffix removed -- as the tag's identity, and enforces two
+invariants whenever a pipeline's tags are written:
+
+* **Variant exclusivity.** A pipeline carries at most one tag per display name.
+  When a list holds several color variants of the same name, the last one wins.
+* **Lexicographic order.** Tags are stored sorted,
+  so the stored order is stable regardless of the order they were supplied in.
```
Comment (mihaibudiu) [inline python/feldera/tags.py]: supplied? By whom?

File: python/feldera/tags.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+"""
+
+import re
+from typing import Iterable, List
+
+# A tag's color suffix: a ``|`` followed by exactly six hex digits at the end of
+# the string (no ``#``, which the backend's tag character set disallows).
+# Recognized only in this exact form, so a name that itself contains ``|``
+# (without a trailing color) keeps its full text.
+_COLOR_SUFFIX = re.compile(r"\|[0-9a-fA-F]{6}$")
```
Comment (mihaibudiu) [inline python/feldera/tags.py]: I thought you didn't allow lowercase hex, I saw a comment about that some place else

File: js-packages/web-console/src/lib/functions/pipelines/tags.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  const name = tagDisplayName(tag)
+  return color ? `${name}|${color.replace(/^#/, '')}` : name
+}
+
+/** Maximum length of a single stored tag, mirroring the backend. */
+export const maximumTagLength = 50
+
+/**
+ * Characters a tag may contain, mirroring the backend's `PATTERN_VALID_TAG`
+ * (crates/pipeline-manager/src/db/types/utils.rs); kept in sync by hand. The
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/functions/pipelines/tags.ts]: whose hand?

File: js-packages/web-console/src/lib/components/common/SlidingPanels.svelte [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+   * Shows one "page" at a time and slides between them, like the panels of a phone
+   * settings screen. `pages` is a fixed, ordered list of named panels and `current`
+   * names the one on screen; changing `current` is the "navigation" that animates
+   * the swap.
+   *
+   * The caller never requests "forward" or "back": navigation is assumed to be linear,
+   * and the slide direction is derived from each page's index in `pages`. Moving to
+   * a larger index the new page enters from the right, moving to a smaller index
+   * slides the page from the left. Comparing indices is a lightweight stand-in
+   * for an actual navigation-history stack, which has not been needed in practice.
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/common/SlidingPanels.svelte]: has not been needed by whom?

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+    tags,
+    knownTags,
+    api
+  }: {
+    /** Name of the pipeline these tags belong to. */
+    pipelineName: string
+    /** Tags currently applied to this pipeline. */
+    tags: string[]
+    /** Every tag known across all pipelines — the pool the popup picks from. */
+    knownTags: Set<string>
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: I find it strange to have both per-pipeline and aggregate info across all pipelines in the same data structure. Is this set duplicated for all pipelines?

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+    /** Pipeline Manager client used to persist tag edits. */
+    api: PipelineManagerApi
+  } = $props()
+
+  // API returns tags already sorted - every client sorts before writing, so a
+  // pipeline's stored tag list is stable regardless of insertion order.
+  // This comparator is used only to
+  // (1) re-establish that order when *writing*edited tags and
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: missing space

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+  // which is built from the unordered cross-pipeline tag pool.
+  const byText = (a: string, b: string) => a.localeCompare(b)
+
+  // Tag edits update the pipeline optimistically and patch the server; the next
+  // list refresh reconciles with the server, which owns `tags`.
+  const { updatePipeline, discardPendingListRefresh } = useUpdatePipelineList()
+  const pipelineList = usePipelineList()
+  const { toastError } = useToast()
+
+  // Sort the edited tags so storage stays sorted (the backend keeps the order),
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: what is "storage"?

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+  // cache that same order optimistically, and return it as the patch body.
+  // This is the one place the client orders tags; readers trust the result.
+  const applyTagsLocally = (name: string, next: string[]) => {
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: what are the tags applied to?
is name a pipeline name? call it "pipeline_name"

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+    const sorted = [...next].sort(byText)
+    updatePipeline(name, (p) => ({ ...p, tags: sorted }))
+    return sorted
+  }
+
+  // Persist a single pipeline's tags. `api.patchPipeline` reports its own failures.
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: reports its own failures -> reports to the users errors that may arise

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+  // Renaming, recoloring, or deleting a tag can touch many pipelines at once.
+  // Update and patch them one batch at a time, awaiting each batch before the
+  // next, so a tag shared by hundreds of pipelines can't fire hundreds of requests
+  // at once. These patches use the bare service call rather than
+  // `api.patchPipeline` so we control how failures surface: within a batch the
+  // first couple of failures toast individually as they land, but the moment a
+  // batch reaches `BATCH_FAILURE_TOAST_LIMIT` failures we give up — let the
+  // current batch finish settling, then raise one toast covering its remaining
+  // failures plus every update in the batches we skip — so a widespread failure
+  // can't bury the user under one toast per pipeline.
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: I think each of these should be a single API call, handled by the server.
Error handling for "paginating" the calls sounds very difficult, especially if multiple users concurrently attempt to do this.

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+      let limitFailureBody = ''
+      await Promise.all(
+        batch.map(({ name, tags }) =>
+          patchPipeline(name, { tags: applyTagsLocally(name, tags) }).catch((error) => {
+            // Toast each failure as it lands, up to the limit. The error that
+            // reaches the limit is carried into the aggregate toast below; later
+            // failures are only counted.
+            failures += 1
+            if (failures < BATCH_FAILURE_TOAST_LIMIT) {
+              toastError('')(new Error(`Failed to update tag “${tagName}” on pipeline “${name}”`))
```
Comment (mihaibudiu) [inline js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: You barely have time to read this.
And if you do, what are you supposed to do?

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte (hunk 11 above)
Comment (mihaibudiu) [reply js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: I think you need some tests for users that concurrently update tags, perhaps in conflicting ways (delete/change color)

**Review by mythical-fred — COMMENTED** [review]
Re-review on 5da0ce4cc8 (since prior COMMENT on 24b27f27d5). The PR has grown well: real backend schema work (V34 migration introducing `client_metadata` JSON, ClientMetadata Rust type, listen_table integration, pipeline_parsing rewrite), a self-contained tag color encoding utility (`tags.ts`), a full picker UI (`Tags.svelte`), a generic `SlidingPanels` component, batched cross-pipeline patching with bounded toast spam, **and** parallel Python SDK helpers (`python/feldera/tags.py`) and Python tests. The structure is now a real feature, not a sketch.

But there are real concerns. The PR is touching three clients — web console, fda (#6530), Python SDK — each with its own copy of the in-band `name|rrggbb` encoding, and the backend has been altered to add a `client_metadata` JSON column. Both of those decisions need scrutiny before this lands.

### Architectural: in-band color encoding, copied three ways

`js-packages/web-console/.../tags.ts`, `crates/fda/src/tags.rs` (in #6530), and `python/feldera/tags.py` each independently re-implement the convention "tag is `name` optionally followed by `|rrggbb`." Three copies of:
- the suffix regex / string slicing,
- the variant-exclusivity rule (one color per name per pipeline),
- the lexicographic-sort invariant,
- the "borrow a color from existing pipelines" rule.

This is the worst version of "shared informal protocol." When any one client changes the rule (say, allows `|RRGGBBAA` for transparency), the other two silently disagree, the backend can't tell, and the cluster ends up with three different views of the same data.

Concrete options:
- **Best: server-side tag color.** A `tag` becomes a structured `{ name, color }` on the wire. Backend can validate, index, and present it. This is a one-table-extension migration — much smaller blast radius than the `client_metadata` JSON column. The argument against ("backend doesn't care about color") doesn't hold up when three clients disagree on the encoding.
- **Acceptable: one shared specification.** A `feldera-tag-format` crate (Rust + WASM bindings for the web) plus a Python wheel that depends on it, with one source of truth for `parseTag` / `tagDisplayName` / `tagWithColor` / `validateTag`. The cost is real (build setup, versioning), but the alternative is three sources of truth that drift.
- **Bare minimum: a TEST that all three implementations agree.** A CI job that runs the same fixture list (`["prod", "prod|ef4444", "a|b", "a|b|ef4444", "release|abcdef", ...]`) through each implementation and asserts they all produce identical `(name, color)` decompositions. This catches drift but does not prevent it.

This decision belongs to mihaibudiu / abhizer (the requested reviewers) since it's an API direction call; flagging here for visibility.

### Architectural: `client_metadata` is a freeform JSON dump

V34__pipeline_metadata.sql introduces a `client_metadata TEXT NOT NULL DEFAULT ''` column holding `json_build_object('description', description)`. The migration drops the standalone `description` column. Reading `pipeline.rs` (+333/-3) and `utils.rs` (+169/-3) the shape is "Rust struct ClientMetadata is the source of truth, schema-free extension is a feature." There are several real risks here:

- **No DB-level shape constraint.** A buggy client or a manual `UPDATE` can write any JSON; downstream readers must trust the `serde_json::from_str` parse. Power-of-Ten #5 (assertions / runtime checks): assert the parse succeeded on every read and fail loudly with which pipeline has bad data.
- **Migration loses validation invariants.** The comment says "existing (migrated) values are exempt until changed." Fine, but the moment a user touches a description with a colon in it on a pipeline created before V34, validation fires. That's user-visible surprise. Worth a one-line warning toast / CLI message ("This description was migrated from an older schema; saving will apply current validation rules") rather than just an opaque error.
- **`TEXT NOT NULL DEFAULT ''`** — empty string is a sentinel meaning "no metadata," parsed as `{}`. Why not `JSONB` with `DEFAULT '{}'::jsonb`? JSONB lets postgres index/filter on metadata fields later (e.g., `WHERE client_metadata @> '{"tags": ["prod"]}'`), which the future "find pipelines by tag" endpoint will want. Switching to JSONB is a one-line schema change now and a major migration later. Make this call before the column ships.
- **The `description` move into `client_metadata`** combined with the "no DB migration needed for new fields" framing reads as "we'd like to dump anything client-side into one column." That's the kind of design that calcifies into "the metadata blob" and becomes a forever-headache for queries. If `tags` and `description` are the only two fields you'll ever put there, make them columns. If they're not, JSONB + a typed Rust struct + a CHECK constraint on the JSON shape.

### Code-level: Tags.svelte batched cross-pipeline patching

`setManyPipelineTags` is much better than the prior version. `PATCH_BATCH_SIZE = 20`, `BATCH_FAILURE_TOAST_LIMIT = 3`, and the aggregate-after-limit toast is well thought through. A few things:

- **`Promise.all` on `patchPipeline(...).catch(...)`.** Good — one rejection won't abort the batch (`Promise.all` resolves because every promise resolves). Pinning this with a comment ("Each `.catch` neutralizes the rejection so `Promise.all` resolves when every patch settles, success or failure") would help the next reader.
- **The optimistic `applyTagsLocally` runs unconditionally.** If the API call fails, the optimistic state stays in `updatePipeline`. The next list refresh reconciles, but until then the user sees the wrong state. Worth either (a) reverting `updatePipeline` on failure, or (b) the toast wording acknowledging "the change will be reverted on the next refresh." Right now the optimistic write looks permanent to the user even when the patch failed.
- **`discardPendingListRefresh()`** is called at the start of `setManyPipelineTags` and again at the start of every `setTags`. Why discard the *pending* refresh rather than queueing one *after* the batch completes? A user editing tag colors rapidly will keep discarding refreshes, and if a patch midway fails silently, the user never sees the divergence. Consider scheduling a final refresh in a `finally` block after the batch loop.
- **`submitEdit` migration logic.** The comment "drop the old tag (and any pre-existing copy of the new one) before adding the new tag, so a pipeline that already had both never ends up with the new tag twice" is great. Worth a test (`Tags.svelte.spec.ts` has 121 lines — does this case appear?) for "pipeline carries both `prod|ef4444` AND `prod|22c55e`, rename to `production|22c55e` collapses to one." Edge case but exactly the kind of thing that breaks silently.

### Code-level: `tags.ts` and parity with `tags.rs` / `tags.py`

- **`validateTag` is duplicative with backend `PATTERN_VALID_TAG`.** The comment says "kept in sync by hand" — that's the drift hazard above. A WASM-compiled validator from the Rust source is one fix; a CI parity test is the minimum.
- **`tagPattern`** allows `\` in tag names. Is that intentional? Backslash in a CLI argument is a UX problem (shell escapes). Worth confirming `fda`'s parser handles `--tag 'a\b'` without surprise.
- **`tagColorOf` returns `defaultTagColor` (`#9ca3af`) for uncolored tags.** Means a "no color" tag is visually indistinguishable from a "gray" tag (palette doesn't include `#9ca3af`, so it's safe today, but the moment someone adds gray to the palette this breaks). Suggest reserving the default gray as a non-pickable sentinel and documenting in `tagColorPalette` that palette entries must not equal `defaultTagColor`.

### Code-level: SlidingPanels

73 lines for a generic sliding-page panel container. Stand-alone is good. Worth a one-line note on whether it's intended for reuse outside the tag picker (the file lives in `common/`) — if yes, a usage example in the JSDoc helps; if no, it should live next to `Tags.svelte` instead.

### Code-level: Python `tags.py`

Same shape as the Rust and TS versions, same in-band encoding. The Python doc string acknowledges "two color variants are considered a tag collision which can but should not occur" — that's an honest acknowledgement that the in-band scheme is leaky. The Python SDK doesn't have a UI, so the "borrow a color" semantics aren't relevant there; it just preserves whatever the caller passed. Reasonable.

### Power-of-Ten

- `setManyPipelineTags` loop is bounded by `updates.length / PATCH_BATCH_SIZE` and gives up after `BATCH_FAILURE_TOAST_LIMIT`. Good.
- `editingTagUsageCount` walks all pipelines on every render — could be expensive on a 10k-pipeline deployment. Memoize on `editingTag` change.
- `[...knownTags]` is taken on every `selectedTags`/`unselectedTags` re-derivation. With `knownTags` of a few thousand, this is fine; if it grows much larger consider a `$derived.by` that caches sorted results.

### Verdict

Direction is right and the polish is real. But the three-way encoding duplication and the `client_metadata` JSON-column architecture are decisions that benefit from explicit sign-off from mihaibudiu / abhizer before they land — both are hard to walk back once shipped. The local code-quality concerns (revert-on-failure, refresh scheduling, edge-case test for color-variant collapse, JSONB vs TEXT) are real but addressable.

COMMENT for now; defer the architectural calls upstream.

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte (hunk 12 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: In the case this occurs you end up with some pipelines with the re-named or re-colored tag and some with the old tag, so you would have to manually repeat the process for the old tag to update the remaining pipelines

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte (hunk 11 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: It indeed is non - trivial. After consulting with Simon we decided to not add any centralized endpoints for managing tags; ignoring the re-naming and re-coloring functionality, there is still a use case for removing a tag from all pipelines it was assigned to - so a way to update multiple pipelines at once is necessary.

I'll discuss it with Simon as it sounds like a general enough feature (PATCH a list of pipelines at once), and I'd rather do that as a part of this PR than merge the batching + concurrent changes tests.

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte (hunk 6 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: There is only a single instance of this set passed here and shared across all instances of Tags.svelte UI component. Each UI component instance manages it's own instance of the popup dialog that offers the user to select from the list of `knownTags`, but since only one popup is visible at a time there is virtually no overhead to this.

File: python/feldera/rest/feldera_client.py (hunk 1 above)
Comment (Karakatiza666) [reply python/feldera/rest/feldera_client.py]: There is already a dedicated `tags` field - part of ClientMetadata in the PipelineInfo, stored by pipeline-manager in the DB. Here it is `stored` because that's how a list of tags is stored in the DB. I'll adjust the comment to clarify it

File: python/feldera/tags.py (hunk 3 above)
Comment (Karakatiza666) [reply python/feldera/tags.py]: I'll tighten this across both clients to only allow lowercase hex

File: js-packages/web-console/src/lib/functions/pipelines/tags.ts (hunk 4 above)
Comment (Karakatiza666) [reply js-packages/web-console/src/lib/functions/pipelines/tags.ts]: I'll rephrase

**Comment by snkas** [comment]
The tags UI looks good, seems convenient adding/removing tags of a pipeline. Same that for now keeping it without color selection. It can be revised in the future if tags really need to be managed as its own object.

As we discussed, I think for now just having the client (Web Console) needing to do multiple requests to add/remove a tag across all pipelines is OK, and then a follow-up PR could add a dedicated endpoint to do that for many if it proves to be too slow.

**Review by mythical-fred — APPROVED** [review]
APPROVE.

The reworked tag-patching flow is the right shape. A few observations worth recording:

The new `promisePool` is clean and the docstring is genuinely helpful — it calls out the "concurrency, not parallelism" point and the `Promise.all`-style rejection semantics up front, which is exactly the kind of contract callers need to know. The `next++` "atomic between awaits" comment is precise: the post-increment runs synchronously before the worker suspends on `await fn(...)`, so no two workers can claim the same index. The spec covers ordering, in-flight cap, sliding behaviour, empty input, the `concurrency >= 1` clamp, and the early-rejection-claims-no-more-items invariant. That last one is the trickiest property and the test for it is well-constructed.

`Tags.svelte` keeps the prior failure semantics intact: failures 1..N-1 toast individually, failure N saves the aggregate body and flips the `stopped` flag, and the final aggregate toast covers failures from N onward plus everything subsequently marked `'skipped'`. The internal try/catch in `fn` keeps the rejection contract of `promisePool` quiet, which matches the docstring. One small thing: once `stopped` is set, remaining workers still burn through one microtask per item to mark them `'skipped'` — fine, it drains in O(n) microtasks, just worth knowing if `updates` ever grows huge.

Minor: `promisePool([], n, ...)` computes `workerCount = Math.max(1, Math.min(n, 0)) = 1` and spawns one worker that exits immediately. Harmless but slightly wasted; an early `if (items.length === 0) return []` would document the intent better and avoid the bookkeeping. Not blocking.

The `SlidingPanels` comment passes my "would a new contributor get the model in 30 seconds" test — explicitly naming the direction-from-index trick as a stand-in for a navigation-history stack is the kind of comment I want more of in this codebase.

Nothing else from me. Ship it.

File: js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte (hunk 6 above)
Comment (mihaibudiu) [reply js-packages/web-console/src/lib/components/pipelines/table/Tags.svelte]: is it readonly? Can you use readonly?

**Review by mihaibudiu — APPROVED** [review]

**Comment by Karakatiza666** [comment]
> is it readonly? Can you use readonly?

Yes and yes, I just usually don't bother using the `readonly` keyword

**Comment by mihaibudiu** [comment]
> > is it readonly? Can you use readonly?
> 
> Yes and yes, I just usually don't bother using the `readonly` keyword

Can you use `Readonly<T>`, `const` or `freeze`? This would catch bugs and help readers

File: python/feldera/tags.py
Comment (abhizer) [inline python/feldera/tags.py:1]: This makes it appear as a module in the python SDK, I'd probably rename the file `_tags.py` so that it is hidden, or move these methods to _helpers.py.

**Review by abhizer — APPROVED** [review]

File: python/feldera/tags.py
Comment (Karakatiza666) [reply python/feldera/tags.py:1]: We want to make at least feldera.tags.tag_display_name public, so I only renamed `normalize_tags` to `_normalize_tags`

**Comment by snkas** [comment]
As I mentioned before, I don't think it is a good idea to have colors get added on as part of the tag as a workaround. If tags need to be more complicated objects, then best way is to just make that change at API/DB level. This will lead to inconsistencies in grouping by tag depending on the client.

Colorless tags is a good first step to validate whether tags are indeed as useful as a concept, and then we can iterate more.

**Review by mythical-fred — APPROVED** [review]
LGTM
